Services.Test/CustomDeviceModelsTest.cs

```diff
@@ -0,0 +1,188 @@
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                null));
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CustomDeviceModelsCanBeUpserted()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { ETag = "oldEtag" };
+            var updatedDeviceModel = new DeviceModel  { ETag = "newETag" };
+
+            this.SaveDeviceModelInStorage(deviceModel);
+            this.UpdateDeviceModelInStorage(updatedDeviceModel);
+
+            // Act
+            this.target.UpsertAsync(deviceModel).Wait();
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                "oldEtag"));
+
+            Assert.Equal("newETag", deviceModel.ETag);
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CreateNewCustomDeviceModelsIfNotExisted()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { Id = "id", ETag = "Etag" };
+
+            this.ThereIsNoCustomDeviceModelsByIdInTheStorage(deviceModel.Id);
+            this.UpdateDeviceModelInStorage(deviceModel);
+
+            // Act
+            this.target.UpsertAsync(deviceModel).Wait();
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                null));
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CustomDeviceModelsCanNotBeUpsertedIfEtagNotMatch()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { ETag = "oldEtag" };
+            var updatedDeviceModel = new DeviceModel { ETag = "newETag" };
+
+            this.SaveDeviceModelInStorage(deviceModel);
+            this.UpdateDeviceModelInStorage(updatedDeviceModel);
+
+            // Act + Assert
+            var modifiedEtagModel = new DeviceModel { ETag = "Non-exist-Etag" };
+            Assert.ThrowsAsync<ConflictingResourceException>(() => this.target.UpsertAsync(modifiedEtagModel));
+        }
+
+        private void SaveDeviceModelInStorage(DeviceModel deviceModel)
+        {
+            var result = new ValueApiModel
+            {
+                ETag = deviceModel.ETag,
+                Data = JsonConvert.SerializeObject(deviceModel)
+            };
+
+            this.storage
+                .Setup(x => x.GetAsync(STORAGE_COLLECTION, It.IsAny<string>()))
+                .ReturnsAsync(result);
+        }
+
+        private void ThereAreNoCustomDeviceModelsInTheStorage()
+        {
+            this.storage
+                .Setup(x => x.GetAllAsync(STORAGE_COLLECTION))
+                .ReturnsAsync(new ValueListApiModel());
+        }
+
+        private void ThereIsNoCustomDeviceModelsByIdInTheStorage(string id)
```

Comment:
pluralization nit: change to "ThereAreNoCustomDeviceModelsByIdInTheStorage" or "ThereIsNoCustomDeviceModelByIdInTheStorage"